config/build.shared.ts

```diff
@@ -0,0 +1,83 @@
+// @ts-check
+
+import crypto from 'crypto';
+import fs from 'fs';
+import type { UserConfig } from 'vitest/config';
+import { resolvePath } from '../tools/shared';
+
+export const external = [
+    'body-parser',
+    'crypto',
+    'express',
+    'fs',
+    'libxslt',
+    'libxmljs',
+    'module',
+    'node:*',
+    'node1-libxmljsmt-myh',
+    'path',
+    'playwright',
+    'perf_hooks',
+    'undici',
+    'url',
+    'vite-node',
+    'vite',
+    'vitest',
+];
+
+export const readFile = (path: string) =>
+    fs.readFileSync(resolvePath(path), 'utf-8');
+
+export const readJSON = (path: string) => {
+    const json = readFile(path);
+
+    return JSON.parse(json);
+};
+
+export const config = readJSON('./config/config.json');
+
+const { version } = readJSON('./package.json');
+
+export const PACKAGE_VERSION = version;
+
+const xslForm = readFile('./src/xsl/openrosa2html5form.xsl');
+const xslModel = readFile('./src/xsl/openrosa2xmlmodel.xsl');
+
+const md5 = (message: string | Buffer) => {
+    const hash = crypto.createHash('md5');
+    hash.update(message);
+
+    return hash.digest('hex');
+};
+
+const HASHED_VERSION = md5(`${xslForm}${xslModel}${PACKAGE_VERSION}`);
+
+export const SERVER_PORT = config.port;
+
+export const define = {
+    [`PACKAGE${'_'}VERSION`]: JSON.stringify(PACKAGE_VERSION),
+    [`HASHED${'_'}VERSION`]: JSON.stringify(HASHED_VERSION),
+    [`SERVER${'_'}PORT`]: JSON.stringify(SERVER_PORT),
+};
+
+export const baseConfig = {
+    assetsInclude: ['**/*.xml', '**/*.xsl'],
+    build: {
+        minify: false,
+        outDir: 'dist',
+        rollupOptions: {
+            external,
+        },
+        sourcemap: true,
+    },
+    define,
+    esbuild: {
+        define,
+        minifyIdentifiers: false,
+        minifySyntax: false,
+        minifyWhitespace: false,
+    },
+    optimizeDeps: {
+        exclude: external,
+    },
+} satisfies UserConfig;
```

Comment:
This line is the primary reason that I went to the effort of converting this module to TypeScript and adding the `vite-node` runner for `web-test-runner.config.ts`. The [`satisfies` operator](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-4-9.html#the-satisfies-operator) is a type-safe way of saying that a value is the intersection of its inferred type and the specified type, e.g.:

```ts
interface Foo {
  a: boolean;
}

const bar = {
  a: true, // Will error if this does not match the `a` type in `Foo`
  b: 1,
} satisfies Foo; // type is { a: boolean; b: number; }
```

This wasn't strictly a **necessary** change, but it helped quite a lot in making sure the various configs remained correct as I added them and addressed build issues found during development. I'm looking forward to using this feature more, as it also overcomes many issues I encounter trying to use the similar [`as const` assertions](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-4.html#const-assertions).